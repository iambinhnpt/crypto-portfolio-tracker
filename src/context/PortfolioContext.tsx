import React, { createContext, useContext, useEffect, useState } from "react";
import { useFetchCryptoList } from "../hooks/useFetchCryptoList";

// Define the context type
interface PortfolioContextType {
    cryptosInfo: any[];
    cryptosInvestmentPortfolio: any[],
    isLoading: boolean;
    updateCryptosInvestmentPortfolio: (newCryptos: any[]) => void;
}

// Create the context
const PortfolioContext = createContext<PortfolioContextType | undefined>(undefined);

// Create the provider component
export const PortfolioProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const { cryptos: cryptosInfo, isLoading } = useFetchCryptoList();
  const [cryptosInvestmentPortfolio, setCryptosInvestmentPortfolio] = useState(JSON.parse(localStorage.getItem('crypto') || '[]'));

  // Function to update the `cryptos` state
  const updateCryptosInvestmentPortfolio = (newCryptos: any[]) => {
    setCryptosInvestmentPortfolio(newCryptos);
  };

  useEffect(() => {
    localStorage.setItem('crypto', JSON.stringify(cryptosInvestmentPortfolio));

  }, [cryptosInvestmentPortfolio])
  

  return (
    <PortfolioContext.Provider value={{ cryptosInfo, isLoading, updateCryptosInvestmentPortfolio, cryptosInvestmentPortfolio }}>
      {children}
    </PortfolioContext.Provider>
  );
};

// Custom hook to use the Portfolio context
export const usePortfolio = (): PortfolioContextType => {
  const context = useContext(PortfolioContext);
  if (!context) {
    throw new Error("usePortfolio must be used within a PortfolioProvider");
  }
  return context;
};