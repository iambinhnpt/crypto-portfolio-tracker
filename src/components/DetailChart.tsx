import React, { useState } from "react";
import { Line } from "react-chartjs-2";
import { useQuery } from "@tanstack/react-query";
import {
  Chart as ChartJS,
  LineElement,
  PointElement,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  CategoryScale,
} from "chart.js";
import { Card, CardContent, Typography, Select, MenuItem, FormControl, InputLabel, CircularProgress } from "@mui/material";
import { usePortfolio } from "../context/PortfolioContext";
import CryptoDetailCard from "./CryptoDetails";

// Register required Chart.js components
ChartJS.register(
  LineElement,
  PointElement,
  LinearScale,
  Title,
  Tooltip,
  Legend,
  CategoryScale
);

interface DetailChartProps {
  cryptoId: string;
}

// Define the type for fetched data
type ChartData = [number, number][];

const fetchMarketChart = async (cryptoId: string, timeRange: string): Promise<ChartData> => {
  const response = await fetch(
    `https://api.coingecko.com/api/v3/coins/${cryptoId}/market_chart?vs_currency=USD&days=${timeRange}`
  );
  if (!response.ok) throw new Error("Network response was not ok");
  const data = await response.json();
  return data.prices;
};

const DetailChart: React.FC<DetailChartProps> = ({ cryptoId }) => {
  const {cryptosInfo} = usePortfolio();
  const cryptoDetail = cryptosInfo.find(item => item.id === cryptoId);
  const [timeRange, setTimeRange] = useState<string>("1"); // Default to 24 hours

  const { data: chartData = [], isLoading, isError } = useQuery<ChartData>({
    queryKey: ["marketChart", cryptoId, timeRange],
    queryFn: () => fetchMarketChart(cryptoId, timeRange),
    enabled: !!cryptoId, // Only fetch if cryptoId exists
    staleTime: 5 * 60 * 1000, // Cache data for 5 minutes
    retry: 2, // Retry failed requests up to 2 times
    }
  );

  // Prepare data for the Line chart
  const data = {
    labels: (chartData as any[]).map((item) => {
      const date = new Date(item[0]);
      if (timeRange === "1") {
        return date.toLocaleTimeString("en-IN", {
          hour: "2-digit",
          minute: "2-digit",
        });
      } else if (timeRange === "30") {
        return date.toLocaleString("en-IN", {
          day: "numeric",
          month: "short",
          hour: "2-digit",
          minute: "2-digit",
        });
      } else {
        return date.toLocaleDateString("en-IN", {
          day: "numeric",
          month: "short",
        });
      }
    }),
    datasets: [
      {
        label: "Price (USD)",
        data: (chartData as any[]).map((item) => item[1]),
        borderColor: "rgba(75, 192, 192, 1)",
        backgroundColor: "rgba(75, 192, 192, 0.2)",
        borderWidth: 2,
        fill: true,
      },
    ],
  };

  const options = {
    responsive: true,
    plugins: {
      legend: { position: "top" as const },
      tooltip: { enabled: true },
    },
    scales: {
      x: {
        title: { display: true, text: "Time" },
      },
      y: {
        title: { display: true, text: "Price (USD)" },
        beginAtZero: false,
      },
    },
  };

  return (
    <div className="p-4 bg-gray-50">
  {/* Crypto Detail Card */}
  {cryptoDetail && <CryptoDetailCard cryptoDetail={cryptoDetail} />}

  {/* Market Chart Section */}
  <Card className="mt-6 shadow-lg">
    <CardContent>
      <div className="flex flex-col lg:flex-row lg:items-center justify-between mb-4">
        <Typography variant="h6" component="h2" className="font-bold">
          Market Chart
        </Typography>
        <FormControl variant="standard" size="small" className="w-full lg:w-1/4">
          <InputLabel id="timeRangeLabel">Select Time Range</InputLabel>
          <Select
            labelId="timeRangeLabel"
            id="timeRange"
            value={timeRange}
            onChange={(e) => setTimeRange(e.target.value)}
            className="bg-white"
          >
            <MenuItem value="1">24 Hours</MenuItem>
            <MenuItem value="30">30 Days</MenuItem>
            <MenuItem value="90">3 Months</MenuItem>
            <MenuItem value="365">1 Year</MenuItem>
          </Select>
        </FormControl>
      </div>

      {/* Chart or Loading/Error */}
      <div className="mt-6">
        {isLoading ? (
          <div className="flex justify-center items-center h-40">
            <CircularProgress />
          </div>
        ) : isError ? (
          <Typography variant="body1" color="error" className="text-center">
            Error loading chart data. Please try again later.
          </Typography>
        ) : (
          <div className="overflow-x-auto">
            <Line data={data} options={options} />
          </div>
        )}
      </div>
    </CardContent>
  </Card>
</div>
  );
};

export default DetailChart;