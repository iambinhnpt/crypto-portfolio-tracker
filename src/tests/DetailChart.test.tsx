import React from "react";
import { render, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { usePortfolio } from "../context/PortfolioContext";
import "@testing-library/jest-dom";
import DetailChart from "../components/DetailChart";

jest.mock("../context/PortfolioContext", () => ({
  usePortfolio: jest.fn(),
}));

jest.mock("react-chartjs-2", () => ({
  Line: () => <div data-testid="chart"></div>,
}));

const queryClient = new QueryClient();

describe("DetailChart", () => {
  const mockPortfolio = {
    cryptosInfo: [
      {
        id: "bitcoin",
        name: "Bitcoin",
        symbol: "BTC",
        image: "https://example.com/bitcoin.png",
        current_price: 50000,
        market_cap: 900000000,
        total_volume: 3000000,
        circulating_supply: 19000000,
        max_supply: 21000000,
        ath: 65000,
        ath_change_percentage: -23,
        ath_date: "2021-11-10",
        atl: 67.81,
        atl_change_percentage: 73745.34,
        atl_date: "2013-07-06",
        high_24h: 51000,
        low_24h: 49000,
        price_change_24h: 1000,
        price_change_percentage_24h: 2.0,
        market_cap_rank: 1,
      },
    ],
  };

  beforeEach(() => {
    (usePortfolio as jest.Mock).mockReturnValue(mockPortfolio);
  });

  it("renders the component and crypto details card", () => {
    render(
      <QueryClientProvider client={queryClient}>
        <DetailChart cryptoId="bitcoin" />
      </QueryClientProvider>
    );

    expect(screen.getByText("Bitcoin (BTC)")).toBeInTheDocument();
    expect(screen.getByText("Market Chart")).toBeInTheDocument();
  });

  it("displays loading state during API fetch", () => {
    render(
      <QueryClientProvider client={queryClient}>
        <DetailChart cryptoId="bitcoin" />
      </QueryClientProvider>
    );

    expect(screen.getByRole("progressbar")).toBeInTheDocument();
  });


  it("renders chart data correctly after fetching", async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () =>
          Promise.resolve({
            prices: [
              [1634235600000, 50000],
              [1634322000000, 50500],
            ],
          }),
      })
    ) as jest.Mock;

    render(
      <QueryClientProvider client={queryClient}>
        <DetailChart cryptoId="bitcoin" />
      </QueryClientProvider>
    );

    expect(await screen.findByTestId("chart")).toBeInTheDocument();
  });
});