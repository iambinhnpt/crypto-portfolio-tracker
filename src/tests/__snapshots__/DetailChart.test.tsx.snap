// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailChart matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-4 bg-gray-50"
  >
    <div
      class="container mx-auto py-3"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root shadow-lg mx-auto css-1j74uew-MuiPaper-root-MuiCard-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiCardContent-root flex justify-center items-center text-center css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1kezxth-MuiAvatar-root"
          >
            <img
              alt="Bitcoin"
              class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
              src="https://example.com/bitcoin.png"
            />
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 text-center mb-4 css-16xl4zq-MuiTypography-root"
          >
            Bitcoin (BTC)
          </div>
        </div>
        <div
          class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Current Price:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $50,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Market Cap:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $900,000,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Total Volume:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $3,000,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                24h High:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $51,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                24h Low:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $49,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Price Change (24h):
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 text-green-600 css-7ctihp-MuiTypography-root"
              >
                2.00% ($1,000)
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Market Rank:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                1
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Circulating Supply:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                19,000,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                Max Supply:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                21,000,000
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                All-Time High:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $65,000 (-23.00%) on 11/10/2021
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 MuiGrid-grid-xl-3 css-v3c8vo-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 font-semibold css-rizt0-MuiTypography-root"
              >
                All-Time Low:
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-7ctihp-MuiTypography-root"
              >
                $67.81 (73745.34%) on 7/6/2013
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root mt-6 shadow-lg css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <div
          class="flex flex-col lg:flex-row lg:items-center justify-between mb-4"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 font-bold css-1miy0lu-MuiTypography-root"
          >
            Market Chart
          </h2>
          <div
            class="MuiFormControl-root w-full lg:w-1/4 css-1iw3t7y-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard css-1m15yum-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="timeRangeLabel"
            >
              Select Time Range
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall bg-white css-j218zi-MuiInputBase-root-MuiInput-root-MuiSelect-root"
            >
              <div
                aria-controls=":r5:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="timeRangeLabel timeRange"
                class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall css-w4dpiq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                id="timeRange"
                role="combobox"
                tabindex="0"
              >
                24 Hours
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                tabindex="-1"
                value="1"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-1gpuzre-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="mt-6"
        >
          <div
            class="overflow-x-auto"
          >
            <div
              data-testid="chart"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
